File: UnitTests/Samples/LateBreaking/DirectoryServices/IdRefConverter/VB/IdRefConverter/IdRefConverter.vb
' ===================================
'   File:      IdRefConverter.vb
'  
'   Summary:   Demonstrates translating between 
'              SecurityIdentifier and NTAccount objects.
'  
' ---------------------------------------------------------------------
'   This file is part of the Microsoft .NET SDK Code Samples.
'  
'   Copyright (C) Microsoft Corporation.  All rights reserved.
'  
' This source code is intended only as a supplement to Microsoft
' Development Tools and/or on-line documentation.  See these other
' materials for detailed information regarding Microsoft code samples.
'  
' THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
' KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
' IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
' PARTICULAR PURPOSE.
' ===================================

Option Strict On
Option Explicit On


Imports System
Imports System.Security.Permissions
Imports System.Security.Principal

<Assembly: System.Reflection.AssemblyVersion("1.0.0.0")> 
<Assembly: SecurityPermission(SecurityAction.RequestMinimum, Execution:=True)> 
<Assembly: EnvironmentPermission(SecurityAction.RequestMinimum, _
								  Read:="USERNAME;USERDOMAIN")> 



Namespace Microsoft.Samples.DirectoryServices

    Public NotInheritable Class IdRefConverter

        Private Sub New()
        End Sub


        Public Shared Sub Main(ByVal args() As String)
            Dim idRef As IdentityReference = Nothing

            '
            ' Get the parameters
            '
            If args.Length <> 2 OrElse (args(0).ToLower() <> "sid" And _
            							args(0).ToLower() <> "nt") Then
                Console.WriteLine("Usage: IdRefConverter sid <SID> or " + _
                						 "IdRefConverter nt <ntAccount>")

                Console.WriteLine("For example:" + vbCrLf + _
                			      "IdRefConverter nt " + _
                Environment.ExpandEnvironmentVariables("%USERDOMAIN%\%USERNAME%") _
                                 + vbCrLf + "IdRefConverter sid S-1-1-0")

                Return
            End If

            '
            ' Translate the identity reference
            '
            Try
                '
                ' Translate SecurityIdentifier to NTAccount
                '
                If args(0).ToLower() = "sid" Then
                    idRef = New SecurityIdentifier(args(1))
                    idRef = CType(idRef.Translate(GetType(NTAccount)), _
                    									  NTAccount)

                '
                ' Translate NTAccount to SecurityIdentifier
                '
                ElseIf args(0).ToLower() = "nt" Then
                    idRef = New NTAccount(args(1))
                    idRef = CType(idRef.Translate(GetType(SecurityIdentifier)), _
                                                          SecurityIdentifier)
                End If

                '
                ' Display the result
                '
                Console.WriteLine(idRef)
            '
            ' If the translation fails, it will be catched here
            '
            Catch e As IdentityNotMappedException
                Console.WriteLine(e.GetType().Name + ":" + e.Message)
            Catch e As Exception
                Console.WriteLine(vbCrLf + "Unexpected exception occured:" + _
                			vbCrLf + vbTab + e.GetType().Name + ":" + e.Message)
            End Try
        End Sub

    End Class

End Namespace

File: UnitTests/Samples/LateBreaking/DirectoryServices/IdRefConverter/CS/IdRefConverter/IdRefConverter.cs
/*=====================================================================
  File:      IdRefConverter.cs

  Summary:   Demonstrates translating between 
             SecurityIdentifier and NTAccount objects.

---------------------------------------------------------------------
  This file is part of the Microsoft .NET SDK Code Samples.

  Copyright (C) Microsoft Corporation.  All rights reserved.

This source code is intended only as a supplement to Microsoft
Development Tools and/or on-line documentation.  See these other
materials for detailed information regarding Microsoft code samples.

THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
PARTICULAR PURPOSE.
=====================================================================*/


using System;
using System.Security.Permissions;
using System.Security.Principal;

[assembly: System.Reflection.AssemblyVersion("1.0.0.0")]
[assembly: SecurityPermission(SecurityAction.RequestMinimum, Execution = true)]
[assembly: EnvironmentPermission(SecurityAction.RequestMinimum,
                                 Read="USERNAME;USERDOMAIN")]

namespace Microsoft.Samples.DirectoryServices
{

public static class IdRefConverter
{
    public static void Main(string[] args)
    {
        IdentityReference idRef = null;

        //
        // Get the parameters
        //
        if (args.Length != 2 || (args[0].ToLower() != "sid" && 
                                  args[0].ToLower() != "nt"))
        {
            Console.WriteLine("Usage: IdRefConverter sid <SID> or "+
                                     "IdRefConverter nt <ntAccount>");
            
            Console.WriteLine("For example:\r\nIdRefConverter nt " + 
            Environment.ExpandEnvironmentVariables("%USERDOMAIN%\\%USERNAME%") +
                              "\r\nIdRefConverter sid S-1-1-0");
            
            return;
        }

        //
        // Translate the identity reference
        //
        try
        {
            //
            // Translate SecurityIdentifier to NTAccount
            //
            if (args[0].ToLower() == "sid")
            {
                idRef = new SecurityIdentifier(args[1]);
                idRef = (NTAccount)idRef.Translate(typeof(NTAccount));                
            }
            //
            // Translate NTAccount to SecurityIdentifier
            //
            else if (args[0].ToLower() == "nt")
            {
                idRef = new NTAccount(args[1]);
                idRef = (SecurityIdentifier)idRef.Translate
                                                   (typeof(SecurityIdentifier));
            }

            //
            // Display the result
            //
            Console.WriteLine(idRef);
        }
        //
        // If the translation fails, it will be catched here
        //
        catch (IdentityNotMappedException e)
        {
            Console.WriteLine(e.GetType().Name + ":" + e.Message);
        }
        catch (Exception e)
        {
            Console.WriteLine("\r\nUnexpected exception occured:\r\n\t" +
                              e.GetType().Name + ":" + e.Message);
        }
        
    }
}

}